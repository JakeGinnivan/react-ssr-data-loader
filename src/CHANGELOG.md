# CHANGELOG

## vNext

- `data-provider.ts`
    - simplify error message as full error details are now provided to consumers