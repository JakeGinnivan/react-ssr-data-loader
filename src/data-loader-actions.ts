export type SuccessAction = {
    type: 'none' | 'initial-fetch' | 'refresh' | 'page' | 'update'
    success: true
}

export type FailedAction = {
    type: 'initial-fetch' | 'refresh' | 'page' | 'update'
    success: false
    error: string
}

export enum LoaderStatus { // The loader is ________ (the data/resource)
    /**
     * The loader is inactive -- not performing any action
     */
    Idle = 'Idle',

    /**
     * The loader has been instantiated and is fetching the resource for the first time
     */
    Fetching = 'Fetching',

    /**
     * The loader is re-fetching the resource
     */
    Refreshing = 'Refreshing',

    /**
     * The loader is fetching the next page of a resurce
     */
    Paging = 'Paging',

    /**
     * The loader is updating the resource
     */
    Updating = 'Updating'
}

export type LoaderState<TData> = {
    status: LoaderStatus
    lastAction: SuccessAction | FailedAction
    /**
     * Some kind of sentinel value so that the object doesn't become top heavy ???
     */

    data: Data<TData>
}

// @ TODO Should we drop dataFromServerSideRender? How do we model not fetching on client
export type Data<TData> =
    | { hasData: true; data: TData; dataFromServerSideRender: boolean }
    | { hasData: false }

export interface DataKeyMap {
    [dataKey: string]: LoaderState<any>
}

export interface DataLoaderState {
    loadingCount: number
    data: { [resourceType: string]: DataKeyMap }
}

export interface ResourceLoadInfo<TAdditionalParameters, TInternalState> {
    resourceType: string
    resourceId: string
    /** Optional additional parameters required to load resource, i.e paging other */
    resourceLoadParams: TAdditionalParameters
    internalState: TInternalState
}

export const INIT = 'resource-data-loader/INIT'
export interface INIT {
    type: typeof INIT
}

export const REFRESH_DATA = 'resource-data-loader/REFRESH_DATA'
// tslint:disable-next-line:class-name
export interface REFRESH_DATA {
    type: typeof REFRESH_DATA
    meta: ResourceLoadInfo<any, any>
}

export const UPDATE_DATA = 'resource-data-loader/UPDATE_DATA'
// tslint:disable-next-line:class-name
export interface UPDATE_DATA {
    type: typeof UPDATE_DATA
    meta: ResourceLoadInfo<any, any>
}

export const NEXT_PAGE = 'resource-data-loader/NEXT_PAGE'
// tslint:disable-next-line:class-name
export interface NEXT_PAGE {
    type: typeof NEXT_PAGE
    meta: ResourceLoadInfo<any, any>
    payload: { existingData: any }
}

export const LOAD_DATA = 'resource-data-loader/LOAD_DATA'
// tslint:disable-next-line:class-name
export interface LOAD_DATA {
    type: 'resource-data-loader/LOAD_DATA'
    meta: ResourceLoadInfo<any, any>
}

export const LOAD_DATA_COMPLETED = 'resource-data-loader/LOAD_DATA_COMPLETED'
// tslint:disable-next-line:class-name
export interface LOAD_DATA_COMPLETED {
    type: 'resource-data-loader/LOAD_DATA_COMPLETED'
    meta: ResourceLoadInfo<any, any>
    payload: {
        data: any
        dataFromServerSideRender: boolean
    }
}

export const LOAD_DATA_FAILED = 'resource-data-loader/LOAD_DATA_FAILED'
// tslint:disable-next-line:class-name
export interface LOAD_DATA_FAILED {
    type: typeof LOAD_DATA_FAILED
    meta: ResourceLoadInfo<any, any>
    payload: Error | string
}

export const UNLOAD_DATA = 'resource-data-loader/UNLOAD_DATA'
// tslint:disable-next-line:class-name
export interface UNLOAD_DATA {
    type: typeof UNLOAD_DATA
    meta: ResourceLoadInfo<any, any>
}

export type Actions =
    | LOAD_DATA
    | LOAD_DATA_COMPLETED
    | LOAD_DATA_FAILED
    | UNLOAD_DATA
    | REFRESH_DATA
    | NEXT_PAGE
    | INIT
    | UPDATE_DATA
