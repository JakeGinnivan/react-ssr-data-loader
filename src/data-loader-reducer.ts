import {
    LoaderState,
    LoaderStatus,
    ResourceLoadInfo,
    DataLoaderState,
    FailedAction,
    Actions,
    LOAD_DATA,
    NEXT_PAGE,
    REFRESH_DATA,
    LOAD_DATA_COMPLETED,
    LOAD_DATA_FAILED,
    UNLOAD_DATA,
    UPDATE_DATA
} from './data-loader-actions'

const defaultState: LoaderState<any> = {
    result: { hasData: false },
    status: LoaderStatus.Idle,
    lastAction: { type: 'none', success: true }
}

const currentDataOrDefault = (
    meta: ResourceLoadInfo<any, any>,
    state: DataLoaderState
): LoaderState<any> => {
    const resourceTypeData = state.data[meta.resourceType]
    if (!resourceTypeData) {
        return defaultState
    }

    const keyData = resourceTypeData[meta.resourceId]
    if (!keyData) {
        return defaultState
    }

    return keyData
}

const statusMap: { [key: string]: FailedAction['type'] } = {
    [LoaderStatus.Paging]: 'page',
    [LoaderStatus.Refreshing]: 'refresh',
    [LoaderStatus.Updating]: 'update',
    [LoaderStatus.Fetching]: 'initial-fetch'
}

export default (
    state: DataLoaderState = {
        data: {},
        loadingCount: 0
    },
    action: Actions
): DataLoaderState => {
    switch (action.type) {
        case LOAD_DATA: {
            const currentOrDefault = currentDataOrDefault(action.meta, state)
            const loading: LoaderState<any> = {
                status: LoaderStatus.Fetching,
                lastAction: currentOrDefault.lastAction, // TODO Should we always go back to idle?
                result: currentOrDefault.result
            }
            return {
                loadingCount: state.loadingCount + 1,
                data: {
                    ...state.data,
                    [action.meta.resourceType]: {
                        ...state.data[action.meta.resourceType],
                        [action.meta.resourceId]: loading
                    }
                }
            }
        }
        // tslint:disable-next-line:no-switch-case-fall-through
        case NEXT_PAGE: {
            const currentState = currentDataOrDefault(action.meta, state)
            const loading: LoaderState<any> = {
                status: LoaderStatus.Paging,
                lastAction: { type: 'none', success: true },
                result: currentState.result
            }
            return {
                loadingCount: state.loadingCount + 1,
                data: {
                    ...state.data,
                    [action.meta.resourceType]: {
                        ...state.data[action.meta.resourceType],
                        [action.meta.resourceId]: loading
                    }
                }
            }
        }
        // tslint:disable-next-line:no-switch-case-fall-through
        case UPDATE_DATA: {
            const currentState = currentDataOrDefault(action.meta, state)
            const loading: LoaderState<any> = {
                status: LoaderStatus.Updating,
                lastAction: { type: 'none', success: true },
                result: currentState.result
            }
            return {
                loadingCount: state.loadingCount + 1,
                data: {
                    ...state.data,
                    [action.meta.resourceType]: {
                        ...state.data[action.meta.resourceType],
                        [action.meta.resourceId]: loading
                    }
                }
            }
        }
        // tslint:disable-next-line:no-switch-case-fall-through
        case REFRESH_DATA: {
            const currentState = currentDataOrDefault(action.meta, state)
            const loading: LoaderState<any> = {
                status: LoaderStatus.Refreshing,
                lastAction: { type: 'none', success: true },
                result: currentState.result
            }
            return {
                loadingCount: state.loadingCount + 1,
                data: {
                    ...state.data,
                    [action.meta.resourceType]: {
                        ...state.data[action.meta.resourceType],
                        [action.meta.resourceId]: loading
                    }
                }
            }
        }
        // tslint:disable-next-line:no-switch-case-fall-through
        case LOAD_DATA_COMPLETED: {
            const currentState = currentDataOrDefault(action.meta, state)
            const lastAction = statusMap[currentState.status]

            const completed: LoaderState<any> = {
                status: LoaderStatus.Idle,
                lastAction: { type: lastAction, success: true },
                result: {
                    hasData: true,
                    data: action.payload.data,
                    dataFromServerSideRender: action.payload.dataFromServerSideRender
                }
            }
            return {
                loadingCount: state.loadingCount - 1,
                data: {
                    ...state.data,
                    [action.meta.resourceType]: {
                        ...state.data[action.meta.resourceType],
                        [action.meta.resourceId]: completed
                    }
                }
            }
        }
        // tslint:disable-next-line:no-switch-case-fall-through
        case LOAD_DATA_FAILED: {
            const currentState = currentDataOrDefault(action.meta, state)
            const lastAction = statusMap[currentState.status]

            const failed: LoaderState<any> = {
                status: LoaderStatus.Idle,
                lastAction: { type: lastAction, success: false, error: action.payload },
                result: currentState.result
            }
            return {
                // this should always occur alongside a LOAD_DATA_COMPLETED so we decrement loadingCount there
                loadingCount: state.loadingCount,
                data: {
                    ...state.data,
                    [action.meta.resourceType]: {
                        ...state.data[action.meta.resourceType],
                        [action.meta.resourceId]: failed
                    }
                }
            }
        }
        // tslint:disable-next-line:no-switch-case-fall-through
        case UNLOAD_DATA: {
            const newState = { loadingCount: state.loadingCount, data: { ...state.data } }
            const dataType = newState.data[action.meta.resourceType]
            delete dataType[action.meta.resourceId]

            if (Object.keys(dataType).length === 0) {
                delete newState.data[action.meta.resourceType]
            } else {
                newState.data[action.meta.resourceType] = dataType
            }

            return newState
        }

        // tslint:disable-next-line:no-switch-case-fall-through
        default:
            return state
    }
}
