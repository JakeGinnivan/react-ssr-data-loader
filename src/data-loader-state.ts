export interface SuccessAction {
    type: 'none' | 'fetch' | 'refresh'
    success: true
}

export interface FailedAction {
    type: 'fetch' | 'refresh'
    success: false
    error: Error & { dataLoadContext?: string }
}

export enum LoaderStatus { // The loader is ________ (the data/resource)
    /**
     * The loader is inactive -- not performing any action
     */
    Idle = 'Idle',

    /**
     * The loader has been instantiated and is fetching the resource for the first time
     */
    Fetching = 'Fetching',

    /**
     * The loader is re-fetching the resource
     */
    Refreshing = 'Refreshing'
}

export interface LoaderState<TData> {
    status: LoaderStatus
    lastAction: SuccessAction | FailedAction
    /**
     * Some kind of sentinel value so that the object doesn't become top heavy ???
     */

    data: Data<TData>
}

// @ TODO Should we drop dataFromServerSideRender? How do we model not fetching on client
export type Data<TData> =
    | { hasData: true; result: TData; dataFromServerSideRender: boolean }
    | { hasData: false }
