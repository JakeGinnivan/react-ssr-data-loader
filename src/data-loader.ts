import * as React from 'react'
import { Dispatch } from 'redux'
import { connect } from 'react-redux'
import {
    ReduxStoreState, DataTypeMap, LoaderDataState,
    LOAD_DATA, LOAD_DATA_FAILED, LOAD_DATA_COMPLETED,
    UNLOAD_DATA, LOAD_NEXT_DATA, ATTACH_TO_DATA,
    DETACH_FROM_DATA
} from './data-loader.redux'

export interface SuccessLoadedState<T> {
    isCompleted: true
    isLoaded: true
    isLoading: false
    isError: false
    data: T
}
export interface ErrorLoadedState {
    isCompleted: true
    isLoaded: false
    isLoading: false
    isError: true
    errorMessage: string
}
export interface BeforeLoadingState {
    isCompleted: false
    isLoaded: false
    isLoading: false
    isError: false
}
export interface LoadingState {
    isCompleted: false
    isLoaded: false
    isLoading: true
    isError: false
}

export type LoadedState<T> = SuccessLoadedState<T> | BeforeLoadingState| ErrorLoadedState | LoadingState

export interface RenderData<T> {
    (loaderProps: LoadedState<T>): React.ReactElement<any> | null
}
export interface OwnProps<T> {
    dataType: string
    dataKey: string
    loadData: () => Promise<T>
    clientLoadOnly?: boolean
    isServerSideRender: boolean
    renderData: RenderData<T>
}

export interface MappedProps {
    store: DataTypeMap
}
export interface DispatchProps {
    dispatch: Dispatch<ReduxStoreState>
}
export interface Props<T> extends OwnProps<T>, MappedProps, DispatchProps { }

const ssrNeedsData = (state: LoaderDataState) => !state || (!state.completed && !state.loading)
const hasValidData = (state: LoaderDataState) => (
    state && state.completed && !state.failed && (
        state.dataFromServerSideRender || state.attachedComponents > 0
    )
)

export class DataLoader<T> extends React.PureComponent<Props<T>, {}> {
    private _isMounted: boolean

    async componentWillMount(): Promise<void> {
        this._isMounted = true
        const loadedState = this.getLoadedState()

        if (this.props.isServerSideRender && ssrNeedsData(loadedState)) {
            return await this.loadData()
        }
        if (!this.props.isServerSideRender) {
            if (hasValidData(loadedState)) {
                this.props.dispatch<ATTACH_TO_DATA>({
                    type: ATTACH_TO_DATA,
                    meta: this.actionMeta(),
                })
            } else {
                return await this.loadData()
            }
        }
    }

    async componentWillReceiveProps(nextProps: Props<T>) {
        if (
            this.props.dataType !== nextProps.dataType ||
            this.props.dataKey !== nextProps.dataKey ||
            this.props.isServerSideRender !== nextProps.isServerSideRender
        ) {
            this.props.dispatch<LOAD_NEXT_DATA>({
                type: LOAD_NEXT_DATA,
                meta: {
                    current: this.actionMeta(),
                    next: this.actionMeta(nextProps)
                }
            })

            await this.performLoadData()
        }
    }

    componentWillUnmount() {
        this._isMounted = false
        const loadedState = this.getLoadedState()
        if (loadedState.attachedComponents > 1) {
            this.props.dispatch<DETACH_FROM_DATA>({
                type: DETACH_FROM_DATA,
                meta: this.actionMeta(),
            })
        } else {
            this.props.dispatch<UNLOAD_DATA>({
                type: UNLOAD_DATA,
                meta: this.actionMeta(),
            })
        }
    }

    private actionMeta = (props = this.props) => ({
        dataType: props.dataType,
        dataKey: props.dataKey,
        dataFromServerSideRender: props.isServerSideRender
    })

    private loadData = async () => {
        this.props.dispatch<LOAD_DATA>({
            type: LOAD_DATA,
            meta: this.actionMeta(),
        })

        await this.performLoadData()
    }

    private performLoadData = async () => {
        try {
            const data = await this.props.loadData()
            if (!this._isMounted) {
                return
            }

            this.props.dispatch<LOAD_DATA_COMPLETED>({
                type: LOAD_DATA_COMPLETED,
                meta: this.actionMeta(),
                payload: data
            })
        } catch (err) {
            if (!this._isMounted) {
                return
            }

            let payload: string
            if (err instanceof Error) {
                payload = err.message
            } else {
                payload = err ? err.toString() : ''
            }

            this.props.dispatch<LOAD_DATA_FAILED>({
                type: LOAD_DATA_FAILED,
                meta: this.actionMeta(),
                payload: payload
            })
        }
    }

    private getLoadedState = (): LoaderDataState => {
        const dataLookup = this.props.store.data[this.props.dataType]
        if (!dataLookup) {
            return undefined
        }
        return dataLookup[this.props.dataKey]
    }

    private getLoadedProps = (): LoadedState<T> => {
        const loadedState = this.getLoadedState()
        if (!loadedState) {
            return undefined
        }

        if (loadedState.completed && loadedState.failed) {
            return {
                isCompleted: true,
                isLoaded: false,
                isLoading: false,
                isError: true,
                errorMessage: loadedState.error,
            }
        } 
        if (loadedState.completed && loadedState.failed === false) {
            return {
                isCompleted: true,
                isLoaded: true,
                isLoading: false,
                isError: false,
                data: loadedState.data
            }
        }
        return {
            isCompleted: false,
            isLoaded: false,
            isLoading: true,
            isError: false,
        }
    }

    render() {
        const loadedProps = this.getLoadedProps() || {
            isCompleted: false,
            isLoading: false,
            isLoaded: false,
            isError: false
        }

        return this.props.renderData(loadedProps)
    }
}

export function createTypedDataLoader<T>() {
    const ConnectedDataLoader = connect<MappedProps, {}, OwnProps<T>>(
        (state: ReduxStoreState) => ({ store: state.dataLoader })
    )(DataLoader)
    return ConnectedDataLoader
}

export default createTypedDataLoader<any>()
