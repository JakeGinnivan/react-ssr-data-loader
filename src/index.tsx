export { RenderData, LoadedState as RenderProps } from './data-loader'
export { default as DataProvider } from './data-provider'
export { default as DataLoaderResources } from './data-loader-resources'
export { default as IsLoading } from './is-loading'
