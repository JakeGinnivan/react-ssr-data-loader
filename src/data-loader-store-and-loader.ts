import reducer from './data-loader-reducer'
import { DataProviderEvents } from './events'
import { DataLoaderState, ResourceLoadInfo } from './data-loader-state'
import * as Actions from './data-loader-actions'
import objectHash from 'object-hash'

export class DataLoaderStoreAndLoader {
    // We need to track this in two places, one with immediate effect,
    // one tied to reacts lifecycle
    private loadingCount = 0
    private state: DataLoaderState

    // When called, data loader will update
    private registeredDataLoaders: {
        [dataLoaderId: string]: () => void
    }
    constructor(
        private onEvent: (event: DataProviderEvents) => void | Promise<any>,
        initialState: DataLoaderState | undefined,
        private performLoad: (metadata: ResourceLoadInfo<any, any>) => Promise<any> | any,
        public isServerSideRender: boolean
    ) {
        if (initialState) {
            this.state = initialState
        } else {
            this.state = reducer(undefined, { type: Actions.INIT })
            this.raiseEvent({
                type: 'state-changed',
                state: this.state
            })
        }
    }

    attach(componentInstanceId: string, update: () => void): void {
        this.registeredDataLoaders[componentInstanceId] = update
    }

    // Returns true when data needs to be unloaded from redux
    detach(componentInstanceId: string) {
        delete this.registeredDataLoaders[componentInstanceId]
    }

    getDataLoaderState(componentInstanceId: string, resourceType: string, dataLoadParams: object) {
        if (!this.registeredDataLoaders[componentInstanceId]) {
            throw new Error(`Data loader with id ${componentInstanceId}`)
        }

        const paramsObject = {
            resourceType,
            ...dataLoadParams
            // Need to get internal state from somewhere?
        }

        const paramsObjectHash = objectHash(paramsObject)

        const dataCache: any = null // TODO implement this
        const loadingState = dataCache.currentState(paramsObjectHash)
        if (loadingState) {
            return loadingState
        }

        dataCache.performLoad()

        const hasParamsChangedForComponent: any = null
        if (hasParamsChangedForComponent(componentInstanceId, paramsObjectHash)) {
            // cleanup old params cache
        }
    }
}
