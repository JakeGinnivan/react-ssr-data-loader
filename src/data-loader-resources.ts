import * as React from 'react'
import { Props, createTypedDataLoader, ActionContext } from './data-loader'

export type LoadResource<TData, TResourceParameters> = (
    resourceId: string,
    resourceParameters: TResourceParameters,
    existingData: TData
) => Promise<TData> | TData

interface Resources {
    [dataType: string]: LoadResource<any, any>
}

export interface PagedData<Datum> {
    pageNumber: number
    data: Datum[]
}
export type RefreshAction = { refresh: () => void }
export type PageActions = { nextPage: () => void } & RefreshAction
export interface Paging {
    /** Defaults to 0 */
    initialOffset?: number
    /** Overrides pageSize for initial page */
    initialSize?: number
    pageSize: number
    /**
     * If true, new page data gets appended to existing data
     * if false, only the current page will be kept
     * Defaults to true
     */
    keepPreviousPagesData?: boolean
}
export type PageComponentProps = { paging: Paging }
type PageState = { page: number }

/** TAdditionalParameters is the type passed to `additionalLoaderProps` on the DataProvider */
export default class DataLoaderResources<TAdditionalParameters> {
    private resources: Resources = {}

    /**
     * When using parameterised resources you cannot have multiple instances of the returned data loader
     * with the same resourceId.
     */
    registerResource<TData, TResourceParameters>(
        resourceType: string,
        loadResource: LoadResource<TData, TResourceParameters & TAdditionalParameters>
    ): React.ComponentClass<Props<TData, RefreshAction> & TResourceParameters> {
        if (this.resources[resourceType]) {
            throw new Error(`The resource type ${resourceType} has already been registered`)
        }

        type ActionsThis = ActionContext<TData, TResourceParameters, {}>
        const typedDataLoader = createTypedDataLoader<
            TData,
            TResourceParameters,
            {},
            RefreshAction
        >(
            resourceType,
            {},
            {
                refresh(this: ActionsThis) {
                    return this.context.dataLoader.refresh({
                        resourceType,
                        resourceId: this.props.resourceId,
                        resourceLoadParams: this.props,
                        internalState: this.internalState
                    })
                }
            }
        )
        this.resources[resourceType] = loadResource

        return typedDataLoader
    }

    /** Page numbers start at 1 */
    registerPagedResource<TData>(
        resourceType: string,
        loadResource: (resourceId: string, paging: Paging, page: number) => Promise<TData[]>
    ): React.ComponentClass<Props<PagedData<TData>, PageActions> & PageComponentProps> {
        if (this.resources[resourceType]) {
            throw new Error(`The resource type ${resourceType} has already been registered`)
        }

        type ActionsThis = ActionContext<TData, PageComponentProps, PageState>
        const typedDataLoader = createTypedDataLoader<
            PagedData<TData>,
            PageComponentProps,
            PageState,
            PageActions
        >(
            resourceType,
            { page: 1 },
            {
                refresh(this: ActionsThis) {
                    return this.context.dataLoader.refresh({
                        resourceType,
                        resourceId: this.props.resourceId,
                        resourceLoadParams: {
                            paging: { ...this.props.paging, keepPreviousPagesData: false }
                        },
                        internalState: { page: 1 }
                    })
                },
                // Loads next page
                nextPage(this: ActionsThis) {
                    return this.context.dataLoader.nextPage({
                        resourceType,
                        resourceId: this.props.resourceId,
                        resourceLoadParams: {
                            paging: { keepPreviousPagesData: true, ...this.props.paging }
                        },
                        internalState: { page: this.internalState().page + 1 }
                    })
                }
            }
        )

        // This async function performs the loading of the paged data
        // it takes care of passing the correct params to the loadResource function
        // then merging the new data when it comes back
        this.resources[resourceType] = async (
            dataKey,
            pageInfo: PageComponentProps & { page: number },
            existingData: PagedData<TData>
        ): Promise<PagedData<TData>> => {
            const pageNumber = pageInfo && pageInfo.page ? pageInfo.page : 1
            const data = await loadResource(dataKey, pageInfo.paging, pageNumber)
            if (existingData && existingData.data && pageInfo.paging.keepPreviousPagesData) {
                return {
                    pageNumber,
                    data: [...existingData.data, ...data]
                }
            }

            return {
                pageNumber,
                data
            }
        }

        return typedDataLoader
    }

    getResourceLoader(dataType: any): LoadResource<any, any> {
        return this.resources[dataType]
    }
}
