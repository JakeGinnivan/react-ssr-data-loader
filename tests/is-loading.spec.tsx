import * as React from 'react'
import { mount, render, ReactWrapper } from 'enzyme'
import { createStore, combineReducers, applyMiddleware, Store } from 'redux'
import { Provider } from 'react-redux'
import { LoadedState, createTypedDataLoader } from '../src/data-loader'
import { ReduxStoreState, reducer } from '../src/data-loader.redux'
import PromiseCompletionSource from './helpers/promise-completion-source'
import ComponentFixture from './helpers/component-fixture'
import IsLoadingFixture, { Verifier } from './helpers/is-loading-fixture'

let store: Store<ReduxStoreState>

beforeEach(() => {
    const logger = store => next => action => {
        // console.log('dispatching', action)
        try {
            let result = next(action)
            // console.log('next state', JSON.stringify(store.getState()))
            return result
        } catch (err) {
            console.error('REDUX ERROR', err)
            throw err
        }
    }

    // TODO Investigate keys of syntax for combineReducers
    store = createStore(
        combineReducers<ReduxStoreState>({ dataLoader: reducer }),
        applyMiddleware(logger)
    )
})

describe('Is loading component', () => {
    it('should be loading if any data loaders are loading', () => {
        const dataLoaderFixture = new ComponentFixture(store, "testKey", false)
        const sut = new IsLoadingFixture(store)

        const verifier = sut.component.find(Verifier)

        expect(verifier.props()).toMatchSnapshot()
        expect(store.getState()).toMatchSnapshot()
    })

    it('should be loading if one of two have finished loading', async () => {
        const dataLoaderFixture = new ComponentFixture(store, "testKey", false)
        const dataLoaderFixture2 = new ComponentFixture(store, "testKey2", false)
        const sut = new IsLoadingFixture(store)
        await dataLoaderFixture.testDataPromise.resolve({
            result: 'Foo!'
        })

        const verifier = sut.component.find(Verifier)

        expect(verifier.props()).toMatchSnapshot()
        expect(store.getState()).toMatchSnapshot()
    })

    it('should not be loading once all data loaders have completed', async () => {
        const dataLoaderFixture = new ComponentFixture(store, "testKey", false)
        const dataLoaderFixture2 = new ComponentFixture(store, "testKey2", false)
        const sut = new IsLoadingFixture(store)
        await dataLoaderFixture.testDataPromise.resolve({
            result: 'Foo!'
        })
        await dataLoaderFixture2.testDataPromise.resolve({
            result: 'Foo2!'
        })

        const verifier = sut.component.find(Verifier)

        expect(verifier.props()).toMatchSnapshot()
        expect(store.getState()).toMatchSnapshot()
    })
})