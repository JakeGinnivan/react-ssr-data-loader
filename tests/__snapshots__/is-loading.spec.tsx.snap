exports[`Is loading component should be loading if any data loaders are loading 1`] = `
Object {
  "isLoading": true,
  "renderCount": 1,
}
`;

exports[`Is loading component should be loading if any data loaders are loading 2`] = `
Object {
  "dataLoader": Object {
    "data": Object {
      "testDataType": Object {
        "testKey": Object {
          "completed": false,
          "failed": false,
          "loading": true,
          "serverSideRender": false,
        },
      },
    },
    "loadingCount": 1,
  },
}
`;

exports[`Is loading component should be loading if one of two have finished loading 1`] = `
Object {
  "isLoading": true,
  "renderCount": 2,
}
`;

exports[`Is loading component should be loading if one of two have finished loading 2`] = `
Object {
  "dataLoader": Object {
    "data": Object {
      "testDataType": Object {
        "testKey": Object {
          "completed": true,
          "data": Object {
            "result": "Foo!",
          },
          "failed": false,
          "loading": false,
          "serverSideRender": false,
        },
        "testKey2": Object {
          "completed": false,
          "failed": false,
          "loading": true,
          "serverSideRender": false,
        },
      },
    },
    "loadingCount": 1,
  },
}
`;

exports[`Is loading component should not be loading once all data loaders have completed 1`] = `
Object {
  "isLoading": false,
  "renderCount": 3,
}
`;

exports[`Is loading component should not be loading once all data loaders have completed 2`] = `
Object {
  "dataLoader": Object {
    "data": Object {
      "testDataType": Object {
        "testKey": Object {
          "completed": true,
          "data": Object {
            "result": "Foo!",
          },
          "failed": false,
          "loading": false,
          "serverSideRender": false,
        },
        "testKey2": Object {
          "completed": true,
          "data": Object {
            "result": "Foo2!",
          },
          "failed": false,
          "loading": false,
          "serverSideRender": false,
        },
      },
    },
    "loadingCount": 0,
  },
}
`;
